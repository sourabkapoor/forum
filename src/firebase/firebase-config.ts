// Import the functions you need from the SDKs you need

import { initializeApp } from "firebase/app";
import { getStorage } from "firebase/storage";

// TODO: Add SDKs for Firebase products that you want to use

// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration

const firebaseConfig = {
  // apiKey: "AIzaSyCLwXf77NuoR0ESLwAAgez_wS_gM8V4J7M",

  // authDomain: "reddit-clone-23782.firebaseapp.com",

  // projectId: "reddit-clone-23782",

  // storageBucket: "reddit-clone-23782.appspot.com",

  // messagingSenderId: "855149209259",

  // appId: "1:855149209259:web:cb03f2ae83010e6feab9ee",

  // measurementId: "G-9D511367MB",
  apiKey: "AIzaSyCEQgSH4Vfzmbn6IVFOE5fu5DtH0p3pAZ8",

  authDomain: "forum-ovidiu-borze.firebaseapp.com",

  projectId: "forum-ovidiu-borze",

  storageBucket: "forum-ovidiu-borze.appspot.com",

  messagingSenderId: "555287825535",

  appId: "1:555287825535:web:98396be2b79a49d0fcc34a",

  measurementId: "G-71MX5H8V0X"

};

// Initialize Firebase

const app = initializeApp(firebaseConfig);
export const storage = getStorage(app);

export default app;
