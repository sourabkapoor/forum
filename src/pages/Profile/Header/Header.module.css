.header {
  height: 39px;
  background-color: var(--primary-color);
  width: 100%;
  display: flex;
  justify-content: center;
}

.header__tabs {
  display: flex;
  column-gap: 10px;
  height: 40px;
  overflow-y: auto;
}

.header__tab {
  color: var(--secondary-color);
  font-size: 14px;
  font-weight: 500;
  line-height: 18px;
  padding: 0 8px;
  white-space: nowrap;
  text-transform: uppercase;
  height: 100%;
  background-color: transparent;
  border: none;
  display: flex;
  align-items: center;
  cursor: pointer;
}

.header__tab--active {
  box-shadow: inset 0 -2px 0 0 var(--secondary-color);
}
