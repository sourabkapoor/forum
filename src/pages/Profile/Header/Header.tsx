import CSSModules from "react-css-modules";
import styles from "./Header.module.css";

interface Action {
  type: string;
}

interface State {
  posts: boolean;
  comments: boolean;
  upvotes: boolean;
  downvotes: boolean;
  saved: boolean;
}
interface Props {
  dispatch: (action: Action) => void;
  activeSection: State;
}

const Header: React.FC<Props> = ({
  dispatch,
  activeSection: { posts, comments, upvotes, downvotes, saved },
}) => {
  return (
    <div styleName="header">
      <div styleName="header__tabs">
        <button
          styleName={`header__tab ${posts && "header__tab--active"}`}
          onClick={() => dispatch({ type: "POSTS" })}
        >
          Posts
        </button>
        <button
          styleName={`header__tab ${comments && "header__tab--active"}`}
          onClick={() => dispatch({ type: "COMMENTS" })}
        >
          Comments
        </button>
        <button
          styleName={`header__tab ${saved && "header__tab--active"}`}
          onClick={() => dispatch({ type: "SAVED" })}
        >
          Saved
        </button>
        <button
          styleName={`header__tab ${upvotes && "header__tab--active"}`}
          onClick={() => dispatch({ type: "UPVOTES" })}
        >
          Upvoted
        </button>
        <button
          styleName={`header__tab ${downvotes && "header__tab--active"}`}
          onClick={() => dispatch({ type: "DOWNVOTES" })}
        >
          Downvotes
        </button>
      </div>
    </div>
  );
};

export default CSSModules(Header, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
