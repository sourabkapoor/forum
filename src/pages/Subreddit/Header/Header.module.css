.header {
  background-color: #efe9f4;
}
.header__background {
  /* background-color: rgb(3, 84, 146); */
  height: 64px;
  padding: 8px 16px;
}
.header__container {
  height: 100px;
  margin: 0 auto;
  max-width: 984px;
}
.header__content {
  display: flex;
  padding: 0 16px 0 24px;
}
.header__subreddit-picture {
  /* height: 72px; */
  max-width: 100px;
  object-fit: cover;
  font-size: 100px !important;
  color: #c262f9;
  background-color: transparent;
  margin-top: -7px;
  margin-bottom: 6px;
  border-radius: 9999px;
}
.header__title-container {
  padding-left: 16px;
  margin-top: 24px;
  display: inline-flex;
  align-items: flex-start;
  justify-content: space-between;
}

.header__title {
  padding-right: 24px;
}
.header__subreddit-name {
  font-size: 28px;
  display: inline-block;
  font-weight: 700;
  color: var(--secondary-color);
  overflow: hidden;
  padding: 0 2px 4px 0;
  text-overflow: ellipsis;
  width: 100%;
}
.header__subreddit-link {
  font-size: 14px;
  font-weight: 500;
  line-height: 18px;
  color: #818384;
}

.header__buttons {
}

.header__button {
  width: 96px;
  background-color: rgb(72, 72, 72);
  border-radius: 9999px;
  color: var(--primary-color);
  font-size: 14px;
  font-weight: 700;
  letter-spacing: unset;
  line-height: 17px;
  text-transform: unset;
  min-height: 32px;
  min-width: 32px;
  padding: 4px 16px;
  border: none;
}

/* .header__button-hover:hover {
  content: "Leave";
  background-color: var(--secondary-color);
} */

.header__button:hover {
  background-color: var(--secondary-color);
  cursor: pointer;
}
