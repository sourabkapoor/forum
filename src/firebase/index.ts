import { initializeApp } from "firebase/app";
import {
  getAuth,
  onAuthStateChanged,
  GoogleAuthProvider,
  signInWithPopup,
  signOut,
  Auth,
} from "firebase/auth";
import {
  getFirestore,
  collection,
  addDoc,
  query,
  orderBy,
  limit,
  onSnapshot,
  setDoc,
  updateDoc,
  doc,
  serverTimestamp,
} from "firebase/firestore";
import {
  getStorage,
  ref,
  uploadBytesResumable,
  getDownloadURL,
} from "firebase/storage";
import { getMessaging, getToken, onMessage } from "firebase/messaging";
import { getPerformance } from "firebase/performance";
import { getAnalytics } from "firebase/analytics";

import app from "./firebase-config";
import { authStateObserver } from "../redux/store";

export const auth = getAuth(app);

export const db = getFirestore(app);

export async function signIn() {
  const provider = new GoogleAuthProvider();
  await signInWithPopup(auth, provider);
}

export async function signOutUser() {
  await signOut(auth);
}

export function initFirebaseAuth() {
  onAuthStateChanged(auth, authStateObserver);
}

export function getUserName() {
  return auth.currentUser?.displayName;
}

export function isUserSignedIn() {
  return !!auth.currentUser;
}

export function getUserId() {
  return auth.currentUser?.uid;
}

export function getUser() {
  return auth.currentUser;
}

initFirebaseAuth();

const analytics = getAnalytics(app);

