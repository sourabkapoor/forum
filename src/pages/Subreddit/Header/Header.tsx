import styles from "./Header.module.css";
import subredditLogo from "../../../assets/subreddit-logo.svg";
import classNames from "classnames";
import CSSModules from "react-css-modules";
import { useAppSelector } from "../../../hooks/hooks";
import {
  selectCommunityData,
  setCommunityData,
} from "../../../features/subreddit/subredditSlice";
import PeopleOutlinedIcon from '@mui/icons-material/PeopleOutlined';
import {
  addDoc,
  arrayRemove,
  arrayUnion,
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  increment,
  runTransaction,
  setDoc,
  updateDoc,
  writeBatch,
} from "firebase/firestore";
import { db, getUser, getUserId } from "../../../firebase";
import { useEffect, useState } from "react";
import { FaReddit } from "react-icons/fa";
import { selectAuthStatus } from "../../../features/auth/authSlice";

interface Props {
  subredditName: string | undefined;
}

const Header: React.FC<Props> = ({ subredditName }) => {
  const { name, imageURL } = useAppSelector(selectCommunityData);
  const [joinButtonText, setJoinButtonText] = useState("Join");

  const isLoggedIn = useAppSelector(selectAuthStatus);

  async function joinCommunity() {
    if (isLoggedIn) {
      try {
        await runTransaction(db, async (transaction) => {
          const userRef = doc(
            db,
            `users/${getUserId()}/communitySnippets/${subredditName}`
          );
          const communityDocRef = doc(db, "subreddits", name);

          const docData = await transaction.get(communityDocRef);

          if (docData.data()?.creatorId === getUserId()) {
            transaction.set(userRef, {
              communityId: subredditName,
              isModerator: true,
            });
          } else {
            transaction.set(userRef, {
              communityId: subredditName,
              isModerator: false,
            });
          }
          const subredditRef = doc(db, `subreddits/${subredditName}`);

          transaction.set(subredditRef, {
            numberOfMembers: increment(1),
          });
        });

        setJoinButtonText("Joined");
      } catch (error) {
        alert(`Could not join community. Please try again: ${error}`);
      }
    } else {
      alert("SIGN IN TO JOIN COMMUNITY");
    }
  }

  async function leaveCommunity() {
    try {
      const communityRef = doc(
        db,
        `users/${getUserId()}/communitySnippets/${subredditName}`
      );

      await deleteDoc(communityRef);
      setJoinButtonText("Join");

      const subredditRef = doc(db, `subreddits/${subredditName}`);

      await updateDoc(subredditRef, {
        numberOfMembers: increment(-1),
      });
    } catch (error) {
      alert(`Could not leave community. Please try again: ${error}`);
    }
  }

  useEffect(() => {
    async function fetchIfUserJoined() {
      try {
        const userCommunities = collection(
          db,
          `users/${getUserId()}/communitySnippets`
        );

        const communities = await getDocs(userCommunities);

        const community = communities
          .docChanges()
          .find(
            (community) => community.doc.data().communityId === subredditName
          );
        if (community) {
          setJoinButtonText("Joined");
        }
      } catch (error) {
        console.log(`Could not see if user joined: ${error}`);
      }
    }
    fetchIfUserJoined();
  }, [subredditName]);

  return (
    <div styleName="header">
      <div styleName="header__background"></div>
      <div styleName="header__container">
        <div styleName="header__content">
          {imageURL ? (
            <img
              styleName="header__subreddit-picture"
              src={imageURL}
              alt="subreddit logo"
            />
          ) : (
            <PeopleOutlinedIcon styleName="header__subreddit-picture" />
          )}
          <div styleName="header__title-container">
            <div styleName="header__title">
              <h1 styleName="header__subreddit-name">{subredditName}</h1>
              <h2 styleName="header__subreddit-link">r/{subredditName}</h2>
            </div>
            <div styleName="header__buttons">
              <button
                styleName="header__button"
                onClick={
                  joinButtonText === "Join" ? joinCommunity : leaveCommunity
                }
              >
                {joinButtonText}
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CSSModules(Header, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
