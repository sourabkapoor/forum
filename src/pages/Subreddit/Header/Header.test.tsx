import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import Header from "./Header";
import * as firestore from "firebase/firestore";
import StoreProvider from "../../../redux/provider";
import * as firebase from "../../../firebase";

describe("Header", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <StoreProvider>
        <Header subredditName="bob jones" />
      </StoreProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("user can join and leave community", () => {
    render(
      <StoreProvider>
        <Header subredditName="bob jones" />
      </StoreProvider>
    );

    jest.mock("../../../firebase", () => ({
      isUserSignedIn: true,
    }));

    const joinCommunityButton = screen.queryByRole("button", { name: "Join" });

    userEvent.click(joinCommunityButton!);

    expect(joinCommunityButton?.textContent).toEqual("Joined");
  });

  test("user can join community", () => {
    render(
      <StoreProvider>
        <Header subredditName="bob jones" />
      </StoreProvider>
    );
    // const mockArrayUnion = jest.fn();
    // const mockUpdateDoc = jest.fn();
    // const mockDoc = jest.fn();
    // const mockGetUserId = jest.fn();

    // jest.mock("firebase/firestore", () => ({
    //   arrayUnion: mockArrayUnion,
    //   doc: mockDoc,
    //   updateDoc: mockUpdateDoc,
    //   getUserId: mockGetUserId,
    // }));

    const mockIsUserSignedIn = jest.fn(() => true);
    // jest.mock("../../../firebase", () => ({
    //   ...jest.requireActual("../../../firebase"),
    //   isUserSignedIn: true,
    // }));

    const spy = jest.spyOn(firebase, "isUserSignedIn");
    const joinCommunityButton = screen.queryByRole("button", { name: "Join" });

    userEvent.click(joinCommunityButton!);

    // expect(spy).toHaveBeenCalled();

    // expect(mockArrayUnion).toBeCalled();
    // expect(mockUpdateDoc).toBeCalled();
    // expect(mockDoc).toBeCalled();
    // expect(mockGetUserId).toBeCalled();
  });
});
