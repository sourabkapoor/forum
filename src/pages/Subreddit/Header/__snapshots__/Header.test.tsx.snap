// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <div
      class="header__background"
    />
    <div
      class="header__container"
    >
      <div
        class="header__content"
      >
        <img
          alt="default subreddit logo"
          class="header__subreddit-picture"
          src="subreddit-logo.svg"
        />
        <div
          class="header__title-container"
        >
          <div
            class="header__title"
          >
            <h1
              class="header__subreddit-name"
            />
            <h2
              class="header__subreddit-link"
            >
              r/
            </h2>
          </div>
          <div
            class="header__buttons"
          >
            <button
              class="header__button"
            >
              Join
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
